### ISSUE TITLE
Segfault after multiple calls to merger objects

### ISSUE THREAD

**Issue by tsellers-r7** [farsightsec/pymtbl#5]
The code linked below will cause a segfault when using `merger` objects to process files with duplicates in them.  

https://gist.github.com/tsellers-r7/aab8a5fd99da8091308da336b643be25

The crashes can be triggered by multiple loops over `.iteritems()` on one instance or by repeatedly creating new objects and then looping over `.iteritems()`.  In all cases the segfault seems to require the processing of multiple files where at least two of them have matching 'keys' which trigger the merge function.

The code contains 3 tests which can be run independently by uncommenting the call to the relevant function.  The test(s) that cause the segfault will vary depending on the operating environment and pymtbl version.  `test1` and `test2` will segfault on Ubuntu 16.04.01 LTS with the latest version of pymtbl and mtbl from Github.  `test3` will segfault on Ubuntu 14.04 using pymtbl 0.3.0 and mtbl 0.6.0. 

Failure seems to be highly situational and depends on the operating environment.  Sometimes `test1` and `test2` will fail on the first iteration of the loop, other times it will be on loop 2 or 3. Adjusting the contents of the `padding` function, which isn't called anywhere, can impact when this failure occurs.  Another example of this is that in `test3` the number of loops/length of file name will change depending on if the debugging print statement on line 93 is comment out or not.

Steps to reproduce the environment on Ubuntu 16.04.1 LTS:
````bash
sudo apt-get install build-essential liblz4-1 liblz4-dev python git dh-autoreconf pkgconf libsnappy1v5 libsnappy-dev zlib1g-dev

mkdir testing
cd testing

git clone https://github.com/farsightsec/mtbl.git
(cd mtbl && ./autogen.sh && ./configure && make && sudo make install && sudo ldconfig)


sudo apt-get install python-pip
pip install cython

git clone https://github.com/farsightsec/pymtbl.git
(cd pymtbl && sudo python setup.py install)
````

save example as 'example.py' and make it executable

```` bash
mkdir data
rm ./data/*.mtbl & ./example.py
````

Sample output

`````bash
Loop # 1
	Enumerate contents of the 'my_merger' object:
		key: a, value:1

Loop # 2
	Enumerate contents of the 'my_merger' object:
Exception TypeError: 'mtbl.iteritems.__next__ (mtbl.c:3106)() takes no arguments (3 given)' in 'mtbl.merge_func_wrapper' ignored

Loop # 3
	Enumerate contents of the 'my_merger' object:
Exception TypeError: 'mtbl.iteritems.__next__ (mtbl.c:3106)() takes no arguments (3 given)' in 'mtbl.merge_func_wrapper' ignored

Loop # 4
	Enumerate contents of the 'my_merger' object:
Exception TypeError: 'mtbl.iteritems.__next__ (mtbl.c:3106)() takes no arguments (3 given)' in 'mtbl.merge_func_wrapper' ignored

Loop # 5
	Enumerate contents of the 'my_merger' object:
Exception TypeError: 'mtbl.iteritems.__next__ (mtbl.c:3106)() takes no arguments (3 given)' in 'mtbl.merge_func_wrapper' ignored
[1]+  Done                    rm ./data/*.mtbl
Segmentation fault (core dumped)

````

Sample output from `test3` which increments the length of one of the input file's name.

````bash
Testing file len: 46 and name: ./data/123456789_123456789_123456789_1234.mtbl
	DEBUG: Adding mtbl file: ./data/123456789_123456789_123456789_1234.mtbl
	DEBUG: Adding mtbl file: ./data/first_fname.mtbl
Enumerate contents of the 'merger' object:
key: a, value:1

Testing file len: 47 and name: ./data/123456789_123456789_123456789_12345.mtbl
	DEBUG: Adding mtbl file: ./data/123456789_123456789_123456789_12345.mtbl
	DEBUG: Adding mtbl file: ./data/first_fname.mtbl
Enumerate contents of the 'merger' object:
key: a, value:1

Testing file len: 48 and name: ./data/123456789_123456789_123456789_123456.mtbl
	DEBUG: Adding mtbl file: ./data/123456789_123456789_123456789_123456.mtbl
	DEBUG: Adding mtbl file: ./data/first_fname.mtbl
Enumerate contents of the 'merger' object:
key: a, value:1

Testing file len: 49 and name: ./data/123456789_123456789_123456789_1234567.mtbl
	DEBUG: Adding mtbl file: ./data/123456789_123456789_123456789_1234567.mtbl
	DEBUG: Adding mtbl file: ./data/first_fname.mtbl
Enumerate contents of the 'merger' object:
key: a, value:1

Testing file len: 50 and name: ./data/123456789_123456789_123456789_12345678.mtbl
	DEBUG: Adding mtbl file: ./data/123456789_123456789_123456789_12345678.mtbl
	DEBUG: Adding mtbl file: ./data/first_fname.mtbl
Enumerate contents of the 'merger' object:
Segmentation fault (core dumped)
````

**Comment by hstern** [comment]
Which version of libmtbl are you using? Can you send a stack trace from gdb?

**Comment by tsellers-r7** [comment]
Crashes caused by `test1` and `test2` work against pymtbl and mtbl built from current master in Github.  I'll see about getting the stack trace.

**Comment by hstern** [comment]
I am unable to reproduce crashes with test1 and test2 using mtbl 0.8.0 and pynmsg 0.4.0. I am able to reproduce the crash with test3.

**Comment by hstern** [comment]
And, if I change the test case as follows:

```diff
--- example.py.1	2016-10-06 10:30:15.671503317 -0300
+++ example.py	2016-10-06 10:29:48.235503474 -0300
@@ -85,7 +85,11 @@
     def test_enumeration(path):
 
         # Create a merger object
-        merger = mtbl.merger(lambda k,v1,v2: v1) 
+        def cmp_fn(k, v1, v2):
+            print k, v1, v2
+            return v1
+
+        merger = mtbl.merger(cmp_fn) 
 
         # Add the mtbl files
         for mtbl_file in os.listdir(path):
```

It works okay.

**Comment by hstern** [comment]
I believe that the lambda function is being garbage collected.

**Comment by hstern** [comment]
Fixed in #6.

**Comment by tsellers-r7** [comment]
This appears to have addressed all three crashes in my test environment.  Thank you very much for looking at the issue and addressing it so quickly.
